FileStore could panic on Close() while expiration in progress

Trace could look like this:
```
panic: failed to switch state from fileInUse to fileOpened for file "data_stores_063501733/foo/msgs.1.dat", state=5
goroutine 352 [running]:
github.com/nats-io/nats-streaming-server/stores.(*filesManager).unlockFile(0xc42016bd00, 0xc42016bec0)
	/home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:990 +0x193
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).doLockFiles(0xc420178b60, 0xc420082f00, 0x1, 0x0, 0xc420041d60)
	/home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1904 +0x304
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).lockIndexFile(0xc420178b60, 0xc420082f00, 0xc420082f00, 0x4a190e)
	/home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1941 +0x48
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).expireMsgs(0xc420178b60, 0x14d94bb50fd603cf, 0x5f5e100, 0x1)
	/home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2423 +0x2a1
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).backgroundTasks(0xc420178b60)
	/home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2706 +0x6e4
created by github.com/nats-io/nats-streaming-server/stores.(*FileStore).newFileMsgStore
	/home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1804 +0x138f
```

--- a/stores/filestore.go
+++ b/stores/filestore.go
@@ -1901,7 +1901,9 @@ func (ms *FileMsgStore) doLockFiles(fslice *fileSlice, onlyIndexFile bool) error
 	}
 	idxWasOpened, err = ms.fm.lockFile(fslice.idxFile)
 	if err != nil {
-		ms.fm.unlockFile(fslice.file)
+		if !datWasOpened {
+			ms.fm.unlockFile(fslice.file)
+		}
 		return err
 	}
 	if !onlyIndexFile {
@@ -2976,6 +2978,16 @@ func (ms *FileMsgStore) Close() error {
 	}
 
 	ms.closed = true
+
+	// Signal the background tasks go-routine to exit
+	ms.bkgTasksDone <- true
+
+	ms.Unlock()
+
+	// Wait on go routines/timers to finish
+	ms.allDone.Wait()
+
+	ms.Lock()
 	var err error
 	if ms.writeSlice != nil {
 		// Flush current file slice where writes happen
@@ -2994,13 +3006,7 @@ func (ms *FileMsgStore) Close() error {
 			err = util.CloseFile(err, slice.idxFile.handle)
 		}
 	}
-	// Signal the background tasks go-routine to exit
-	ms.bkgTasksDone <- true
-
 	ms.Unlock()
-
-	// Wait on go routines/timers to finish
-	ms.allDone.Wait()
 
 	return err
 }
@@ -3583,6 +3589,12 @@ func (ss *FileSubStore) Close() error {
 			ss.allDone.Done()
 		}
 	}
+	ss.Unlock()
+
+	// Wait on timers/callbacks
+	ss.allDone.Wait()
+
+	ss.Lock()
 	var err error
 	if ss.fm.remove(ss.file) {
 		if ss.file.handle != nil {
@@ -3591,9 +3603,6 @@ func (ss *FileSubStore) Close() error {
 		}
 	}
 	ss.Unlock()
-
-	// Wait on timers/callbacks
-	ss.allDone.Wait()
 
 	return err
 }

--- a/stores/filestore_msg_test.go
+++ b/stores/filestore_msg_test.go
@@ -1412,3 +1412,26 @@ func TestFSMsgStoreBackgroundTaskCrash(t *testing.T) {
 	time.Sleep(50 * time.Millisecond)
 	// It should not have crashed.
 }
+
+func TestFSPanicOnStoreCloseWhileMsgsExpire(t *testing.T) {
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+
+	limits := testDefaultStoreLimits
+	limits.MaxAge = 30 * time.Millisecond
+
+	fs, _, err := newFileStore(t, defaultDataStore, &limits)
+	if err != nil {
+		t.Fatalf("Unable to create store: %v", err)
+	}
+	defer fs.Close()
+
+	cs := storeCreateChannel(t, fs, "foo")
+
+	for i := 0; i < 100; i++ {
+		storeMsg(t, cs, "foo", []byte("msg"))
+	}
+
+	time.Sleep(30 * time.Millisecond)
+	fs.Close()
+}

